File: packages/core/types/template/ProgramStructures.ts
import { IBaseNode, TemplateNodeTypes } from "./BaseNode/BaseTypes";

export interface IArrayDeclaration extends IBaseNode {
  elementType: string;
  length: number | string;
  name: string;
  nodeType: TemplateNodeTypes.ArrayDeclaration;
  storage?: string;
}

export interface IFunctionDeclaration extends IBaseNode {
  nodeType: TemplateNodeTypes.FunctionDeclaration;
}

export interface IFunctionDefinition extends IBaseNode {
  name: string;
  nodeType: TemplateNodeTypes.FunctionDefinition;
  returnType: string;
}

export interface IParameterDeclaration extends IBaseNode {
  nodeType: TemplateNodeTypes.ParameterDeclaration;
  name: string;
  type: string;
  size?: string;
}

export interface IParameterList extends IBaseNode {
  nodeType: TemplateNodeTypes.ParameterList;
}

export interface IPointerDeclaration extends IBaseNode {
  nodeType: TemplateNodeTypes.PointerDeclaration;
  name: string;
  pointingType: string;
  level: number;
}

export interface ITranslationUnit extends IBaseNode {
  nodeType: TemplateNodeTypes.TranslationUnit;
}

export interface IVariableDeclaration extends IBaseNode {
  name: string;
  nodeType: TemplateNodeTypes.VariableDeclaration;
  storage?: string;
  type: string;
}

File: packages/core/src/core/types/template/program_structures.py
"""Program structure node type definitions."""

from typing import Optional

from .BaseNode.base_types import IBaseNode, TemplateNodeTypes


class IArrayDeclaration(IBaseNode):
    """Array declaration node."""

    elementType: str
    length: int | str
    name: str
    nodeType: TemplateNodeTypes = TemplateNodeTypes.ArrayDeclaration
    storage: Optional[str] = None


class IFunctionDeclaration(IBaseNode):
    """Function declaration node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.FunctionDeclaration


class IFunctionDefinition(IBaseNode):
    """Function definition node."""

    name: str
    nodeType: TemplateNodeTypes = TemplateNodeTypes.FunctionDefinition
    returnType: str


class IParameterDeclaration(IBaseNode):
    """Parameter declaration node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.ParameterDeclaration
    name: str
    type: str
    size: Optional[str] = None


class IParameterList(IBaseNode):
    """Parameter list node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.ParameterList


class IPointerDeclaration(IBaseNode):
    """Pointer declaration node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.PointerDeclaration
    name: str
    pointingType: str
    level: int


class ITranslationUnit(IBaseNode):
    """Translation unit node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.TranslationUnit


class IVariableDeclaration(IBaseNode):
    """Variable declaration node."""

    name: str
    nodeType: TemplateNodeTypes = TemplateNodeTypes.VariableDeclaration
    storage: Optional[str] = None
    type: str


